### PR TITLE
Deleting interlinked components breaks the application

### PR DESCRIPTION
**Pull request by sudhirverma**
Fix: #878

### CODE REVIEW

File: src/openshift/component.ts [hunk 1]
```diff
@@ -67,13 +67,24 @@ export class Component extends OpenShiftItem {
...
                 return Component.odo.deleteComponent(component);
             }).then(() => `Component '${name}' successfully deleted`)
                 .catch((err) => Promise.reject(`Failed to delete Component with error '${err}'`));
         }
     }
 
+    static async unlinkComponent(component: OpenShiftObject) {
+        const compData = await Component.odo.execute(Command.getdescribeComponent(component.getParent().getParent().getName(), component.getParent().getName(), component.getName()));
+        const compObj: JSON = JSON.parse(compData.stdout);
+        if (compObj['status'].linkedComponents) {
```
Comment (mohitsuman) [inline src/openshift/component.ts]: @sudhirverma try to use map function here and pass the ket to `unlinkComponentTo`.

**Comment by mohitsuman** [comment]
@sudhirverma please do not do force push, it is tough to review the previous commit comment.

File: src/odo.ts [hunk 2]
```diff
@@ -131,6 +131,9 @@ export class Command {
     static describeComponent(project: string, app: string, component: string) {
         return `odo describe ${component} --app ${app} --project ${project}`;
     }
+    static getdescribeComponent(project: string, app: string, component: string) {
```
Comment (dgolovin) [inline src/odo.ts]: lets name it `describeComponentJson` for consistency.

File: src/odo.ts [hunk 3]
```diff
@@ -131,6 +131,9 @@ export class Command {
...
+        return `odo describe ${component} --app ${app} --project ${project} -o json`;
```
Comment (dgolovin) [inline src/odo.ts]: should reuse `describeComponent' method.
``` JavaScript
return `${Command.describeComponent(project, app. component)} -o json`;
```

File: src/odo.ts [hunk 4]
```diff
@@ -198,6 +201,10 @@ export class Command {
     static getComponentJson(project: string, app: string, component: string) {
         return `oc get service ${component}-${app} --namespace ${project} -o json`;
     }
+
+    static unlinkComponentTo(comp1, comp2) {
```
Comment (dgolovin) [inline src/odo.ts]: Should be named just `unlinkComponents`.

File: src/openshift/component.ts [hunk 5]
```diff
@@ -67,13 +67,25 @@ export class Component extends OpenShiftItem {
...
-            return Progress.execFunctionWithProgress(`Deleting the Component '${component.getName()} '`, () => {
+            return Progress.execFunctionWithProgress(`Deleting the Component '${component.getName()} '`, async () => {
+                await Component.unlinkComponent(component);
                 return Component.odo.deleteComponent(component);
             }).then(() => `Component '${name}' successfully deleted`)
                 .catch((err) => Promise.reject(`Failed to delete Component with error '${err}'`));
         }
     }
 
+    static async unlinkComponent(component: OpenShiftObject) {
```
Comment (dgolovin) [inline src/openshift/component.ts]: Should be named `unlinkAllCOmponents`.

File: src/odo.ts [hunk 6]
```diff
@@ -198,6 +201,10 @@ export class Command {
...
+        return `odo unlink ${comp2} --component ${comp1}`;
```
Comment (dgolovin) [inline src/odo.ts]: `--namespace proj` and `--app app` options are missing.

File: src/openshift/component.ts [hunk 7]
```diff
@@ -67,13 +67,25 @@ export class Component extends OpenShiftItem {
         const name: string = component.getName();
         const value = await window.showWarningMessage(`Do you want to delete Component '${name}\'?`, 'Yes', 'Cancel');
         if (value === 'Yes') {
-            return Progress.execFunctionWithProgress(`Deleting the Component '${component.getName()} '`, () => {
+            return Progress.execFunctionWithProgress(`Deleting the Component '${component.getName()} '`, async () => {
+                await Component.unlinkAllCOmponents(component);
```
Comment (mohitsuman) [inline src/openshift/component.ts]:
```suggestion
                await Component.unlinkAllComponents(component);
```

File: src/openshift/component.ts [hunk 8]
```diff
@@ -67,13 +67,25 @@ export class Component extends OpenShiftItem {
...
                 return Component.odo.deleteComponent(component);
             }).then(() => `Component '${name}' successfully deleted`)
                 .catch((err) => Promise.reject(`Failed to delete Component with error '${err}'`));
         }
     }
 
+    static async unlinkAllCOmponents(component: OpenShiftObject) {
```
Comment (mohitsuman) [inline src/openshift/component.ts]:
```suggestion
    static async unlinkAllComponents(component: OpenShiftObject) {
```

File: test/openshift/component.test.ts [hunk 9]
```diff
@@ -356,9 +356,39 @@ suite('OpenShift/Component', () => {
...
+        test('Should able to unlink the component', async () => {
+            await Component.unlinkAllCOmponents(componentItem);
+            execStub.calledOnce;
+        });
+    });
+
     suite('del', () => {
 
         setup(() => {
+            sandbox.stub(Component, 'unlinkAllCOmponents');
```
Comment (mohitsuman) [inline test/openshift/component.test.ts]:
```suggestion
            sandbox.stub(Component, 'unlinkAllComponents');
```

**Pull request merged**